packages/liveblocks-core/src/lib/__tests__/IncrementalJsonParser.test.ts

```diff
@@ -0,0 +1,404 @@
+    expect(parse('{"key":"value"}   ')).toEqual({ key: "value" });
+    expect(parse('{"a":1,"b":2} \n ')).toEqual({ a: 1, b: 2 });
+    expect(parse('{"nested":{"obj":{}}}  ')).toEqual({ nested: { obj: {} } });
+  });
+
+  test("handles emojis and newlines", () => {
+    // Complete emojis in strings
+    expect(parse('{"message":"Hello 👋 world')).toEqual({
+      message: "Hello 👋 world",
+    });
+    expect(parse('{"reaction":"🎉","status":"complete')).toEqual({
+      reaction: "🎉",
+      status: "complete",
+    });
+
+    // Partial/incomplete emojis (multi-byte sequences)
+    expect(parse('{"partial":"test 👋')).toEqual({ partial: "test 👋" });
+    expect(parse('{"emoji":"🎉🎊')).toEqual({ emoji: "🎉🎊" });
+
+    // Newlines and whitespace in strings
+    expect(parse('{"text":"line1\\nline2')).toEqual({ text: "line1\nline2" });
+    expect(parse('{"multiline":"first line\\nsecond')).toEqual({
+      multiline: "first line\nsecond",
+    });
+
+    // Mixed emojis, newlines, and regular content
+    expect(parse('{"log":"User clicked 👆\\nAction: success ✅')).toEqual({
+      log: "User clicked 👆\nAction: success ✅",
+    });
+
+    // Emojis with arrays and objects
+    expect(parse('{"reactions":["👍","👎","❤️')).toEqual({
+      reactions: ["👍", "👎", "❤️"],
+    });
+    expect(parse('{"user":{"name":"Alice","status":"🟢 online')).toEqual({
+      user: { name: "Alice", status: "🟢 online" },
+    });
+  });
+});
+
+describe("partial keyword recognition", () => {
+  test("recognizes partial null keyword", () => {
+    expect(parse('{"key":n')).toEqual({ key: null });
+    expect(parse('{"key":nu')).toEqual({ key: null });
+    expect(parse('{"key":nul')).toEqual({ key: null });
+  });
+
+  test("recognizes partial true keyword", () => {
+    expect(parse('{"key":t')).toEqual({ key: true });
+    expect(parse('{"key":tr')).toEqual({ key: true });
+    expect(parse('{"key":tru')).toEqual({ key: true });
+  });
+
+  test("recognizes partial false keyword", () => {
+    expect(parse('{"key":f')).toEqual({ key: false });
+    expect(parse('{"key":fa')).toEqual({ key: false });
+    expect(parse('{"key":fal')).toEqual({ key: false });
+    expect(parse('{"key":fals')).toEqual({ key: false });
+  });
+
+  test("recognizes partial keywords in arrays", () => {
+    expect(parse('{"arr":[t')).toEqual({ arr: [true] });
+    expect(parse('{"arr":[n')).toEqual({ arr: [null] });
+    expect(parse('{"arr":[f')).toEqual({ arr: [false] });
+  });
+
+  test("recognizes partial keywords after commas", () => {
+    expect(parse('{"arr":[true,f')).toEqual({ arr: [true, false] });
+    expect(parse('{"a":1,"b":n')).toEqual({ a: 1, b: null });
+  });
+
+  test("does not complete keywords in strings", () => {
+    // This should not complete "n" to "null" because it's inside a string
+    expect(parse('{"key":"n')).toEqual({ key: "n" });
+  });
+
+  test("fixes the original counterexample", () => {
+    // This was the counterexample that revealed the issue
```

Comment:
Could this be more specific? I'm not sure what the original counterexample or the issue are

Reply:
LOL I also don't know anymore.